### ISSUE TITLE
RFC: Deprecate Selectors With Props

### ISSUE THREAD

**Issue by timdeschryver** [ngrx/platform#2980]
# RFC: Deprecate VS Improve Selector With Props
​
I created this to jot down some thoughts about the current state of selectors with props with the intention to come up with a game plan. The result should either be to deprecate selectors with props (and later removed), or to tackle the current problems.
​
## Developer Experience
​
The biggest problem in terms of developer experience is that the types inside the selector aren't inferred correctly
(https://github.com/ngrx/platform/issues/2205). With the release of Angular 12, a new project will be strict by default, so I assume more developers will stumble against this issue.
​
A fix would be to improve the types.
​
## Performance
​
One of the big benefits of using selectors is that a selector is memoized, but with selectors with props it's very easy to "break" this. Most of the problems/questions I see is because selector with props aren't used correctly, resulting in selectors that are executed multiple times. This isn't something a user might expect, and arises some confusion. Even when they're used correctly, shared "child selectors" will be invoked for each time that child selector is used in a selector with props (https://github.com/ngrx/platform/issues/2862).
​
We could improve the docs and try to explain the concept around selector with props better.
Currently, I don't see how we can reduce the number of calls.
​
## Maintenance
​
The NgRx has [a lot of overloads](https://github.com/ngrx/platform/blob/master/modules/store/src/selector.ts#L121-L507) to correctly type selectors and selectors with props. We could refactor this with [Variadic Tuple Types](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-4-0.html#variadic-tuple-types), with the benefit that we can support to compose N amount of selectors. I'm not certain, but I think removing selectors with props, would help to refactor the selector's types.
​
## Advocating for factory selectors
​
As mentioned in the earlier sections, I do think that we can improve and fix the problems with the current state of selectors with props.
​
But I do also think that even if we do this, that selector with props will still be "miss-used" simply because it's a concept that has to learned. I've seen numerous questions of users using selector with props, but are still asking on how to pass those props to child selectors (even though the props are passed down to all selectors).
​
So why not keep things simple with factory selectors to pass props to a selector?
By doing this the intention is clear, and it's something that everyone can understand because these are just functions. Simply for this reason, I encourage using factory selectors.
​
Factory selectors also fix the performance hit, although I assume that the difference is negligible.

The only downside is that selectors aren't automatically passed down to all selectors (I think that the majority don't know this about selectors with props, and thus are also not using it). Instead, we have to pass down the props manually.

## TLDR

Selector with props is a new concept, are hard to understand, and are easy to make "mistakes" with.
Instead, we can create the same behavior with factory selectors, without all the overhead.
A bonus would be that I assume that we can get rid of a lot of types (and a bit of code) in the NgRx repo and replace them with variadic tuple types.

To emphasize, we can probably resolve most of the technical problems, but I fear that selectors with props are harder "to get right".

## Demo
​
```ts
interface ReducerState {
  count: number;
}
​
let logs: string[] = [];
​
const selector = createFeatureSelector<ReducerState>('s');
​
const selectCount = createSelector(selector, (s) => {
  logs.push(`[selectCount] ${s.count}`);
  return s.count;
});
​
const selectCountMultipliedProps = () =>
  // 👎 the result type of the selector is lost (`count`)
  // Parameter 'count' implicitly has an 'any' type
  createSelector(
    selectCount,
    (count: number, props: { multiplier: number }) => {
      logs.push(`[selectCountMultipliedProps] ${count} * ${props.multiplier}`);
      return count * props.multiplier;
    }
  );
​
const selectCountMultiplied = (props: { multiplier: number }) =>
  // 👍 `count` knows that it's a number
  createSelector(selectCount, (count) => {
    logs.push(`[selectCountMultiplied] ${count} * ${props.multiplier}`);
    return count * props.multiplier;
  });
​
beforeEach(() => {
  logs.length = 0;
});
​
it('selector with props', () => {
  const state = { s: { count: 5 } };
​
  const selector2 = selectCountMultipliedProps();
  const selector3 = selectCountMultipliedProps();
​
  selector2(state, { multiplier: 2 });
  selector2(state, { multiplier: 2 });
  selector3(state, { multiplier: 3 });
  selector2(state, { multiplier: 2 });
​
  console.log('selector with props', logs);
​
  // selector with props [
  //   '[selectCount] 5',
  //   '[selectCountMultipliedProps] 5 * 2',
  //   '[selectCount] 5',
  //   '[selectCountMultipliedProps] 5 * 2',
  //   '[selectCount] 5',
  //   '[selectCountMultipliedProps] 5 * 3',
  //   '[selectCount] 5',
  //   '[selectCountMultipliedProps] 5 * 2'
  // ]
});
​
it('selector with same props', () => {
  const state = { s: { count: 5 } };
​
  const selector2 = selectCountMultipliedProps();
  const props2 = { multiplier: 2 };
  const selector3 = selectCountMultipliedProps();
  const props3 = { multiplier: 3 };
​
  selector2(state, props2);
  selector2(state, props2);
  selector3(state, props3);
  selector2(state, props2);
​
  console.log('selector with same props', logs);
​
  // selector with same props [
  //   '[selectCount] 5',
  //   '[selectCountMultipliedProps] 5 * 2',
  //   '[selectCount] 5',
  //   '[selectCountMultipliedProps] 5 * 3'
  // ]
});
​
it('factory selector', () => {
  const state = { s: { count: 5 } };
​
  selectCountMultiplied({ multiplier: 2 })(state);
  selectCountMultiplied({ multiplier: 2 })(state);
  selectCountMultiplied({ multiplier: 3 })(state);
  selectCountMultiplied({ multiplier: 2 })(state);
​
  console.log('factory selector', logs);
​
  // factory selector [
  //   '[selectCount] 5',
  //   '[selectCountMultiplied] 5 * 2',
  //   '[selectCountMultiplied] 5 * 2',
  //   '[selectCountMultiplied] 5 * 3',
  //   '[selectCountMultiplied] 5 * 2'
  // ]
});
​
it('factory selector with same reference', () => {
  const state = { s: { count: 5 } };
​
  const select2 = selectCountMultiplied({ multiplier: 2 });
  const select3 = selectCountMultiplied({ multiplier: 3 });
  select2(state);
  select2(state);
  select3(state);
  select2(state);
​
  console.log('factory selector with same reference', logs);
​
  // factory selector with same reference [
  //   '[selectCount] 5',
  //   '[selectCountMultiplied] 5 * 2',
  //   '[selectCountMultiplied] 5 * 3',
  // ]
});

**Comment by brandonroberts** [comment]
I agree with the rationale to deprecate them and eventually remove them. We initially ported them over for feature parity with Reselect. We'll have to make it more clear on how to test selectors created by a factory function also.

**Comment by mistic100** [comment]
I fail to understand what we could use instead by reading the example code.

I have a huge code base using selectors with props for rights management.

```ts
export const hasRight = createSelector(selectUser, (user: UserDTO, { rightCode }: RightDTO) => {
    return user?.rights?.indexOf(rightCode) !== -1;
});


const canEditClient$ = this.store.select(hasRight, RIGHT_EDIT_CLIENT);
```

What will be the recommended way to achieve this ?

**Comment by mmulhearn** [comment]
@mistic100 I may be wrong, but I don't think what you're doing needs to be an observable.  The user, which I assume is the logged in user, is constant and won't emit changes so making that an observable seems like overkill. Then, once that is no longer an observable, then you can create utility functions to determine if the user has X right, which is a simple `indexOf` like you h ave against a non-observable.  You're really using the selector to get around the difficulty created by making your logged in user a selector/observable.

**Comment by timdeschryver** [comment]
@mistic100 the selector needs to be rewritten as a "factory selector".

```ts
export const hasRight = (rightCode: RightCode) => createSelector(selectUser, (user) => {
    return user?.rights?.indexOf(rightCode) !== -1;
});

// you can consume it as
const canEditClient$ = this.store.select(hasRight(RIGHT_EDIT_CLIENT));
```

**Comment by mistic100** [comment]
@mmulhearn But the user is loaded via effects and stored in the Store, and because NgRx don't expose synchronous access to the store, it **has** to be an Observable.

@timdeschryver Thank you, that's a transformation that can be automated

**Comment by mmulhearn** [comment]
@mistic100 word, yeah. i showed some people on my team my comment and yours and it started a large discussion of "what is state and how are we using NGRX in relation to static, unchanging data" and that resulted in my comment being completely wrong lol. glad @timdeschryver had an answer as we'll need to update some of our selectors as well.

**Comment by mistic100** [comment]
@mmulhearn well I wouldn't say it's wrong. If I had to redo the state management in my company I would certainly consider other alternatives to those "static" data. But we are off-topic :-)

**Comment by bbonczek** [comment]
@timdeschryver looks good. I am wondering - is it going to make #2176 no longer valid? Are we going to be able to mock response of these selectors based on given payload if we use these "factory selectors"?

**Comment by timdeschryver** [comment]
@bbonczek since it's a factory function you should be able to mock (with jasmine, jest, sinon, ...) the selector, although not with `overrideSelector`.

**Comment by wrslatz** [comment]
Will there be any guidance published on how to override factory selectors in testing? `overrideSelector` no longer works in this case (that I can figure out, at least)

**Comment by mmulhearn** [comment]
> Will there be any guidance published on how to override factory selectors in testing? `overrideSelector` no longer works in this case (that I can figure out, at least)

I found you need to overrideSelector the lowest factory selector as they daisy chain.

**Comment by wrslatz** [comment]
> I found you need to overrideSelector the lowest factory selector as they daisy chain.

This is the approach I'm going with for now, but it requires knowing the implementation of the factory selector (knowing what selector it calls and how it uses the result). Not terrible, but not as clean as it was before.

**Comment by mmulhearn** [comment]
Yup, you're right.  I've tried using `jasmine.createSpy` but have been unsuccessful in doing so.  This is the only way I've gotten it to work.  If you figure out another way, please lmk!

**Comment by timdeschryver** [comment]
@theefer I get and hear what you're saying, and these are valid concerns.

Depending on how you structure the selectors and pass the to the selectors, the current API would also invoke the `lookup` and `denormalize` methods multiple times. In the above snippet, the prop `{collection}` would still have triggered those methods more than once, because these are 3 different instances of props.

You're right about inheriting props, but we felt like this was a feature that wasn't commonly known/used and lead to unwanted scenarios. For us, the benefits of removing props outweigh the cost of having props that were passed down to child selectos. 

You're also right that each selector will work with its own cache, and for most of the cases this is fine (and desired). If the project that you're working on has performance issues because of this change, I would be interested in having a reproduction to see where the problem lies and how we can improve that (either in this repo, or in your codebase).

A potential fix (and this is how I use selectors) would be to just create one "viewmodel" selector, instead of using multiple selectors in a component. You could also look into adding a [memoization layer in the projector method of a selector](https://timdeschryver.dev/blog/parameterized-selectors).

**Comment by fgirardey** [comment]
> I would be interested in having a reproduction to see where the problem lies and how we can improve that (either in this repo, or in your codebase).

I might have a reproduction : https://stackblitz.com/edit/angular-ngrx-selectors-bad?file=src/app/hello.component.ts

In my example I use `ngrxPush` to trigger render (zoneless example) and since the selector factory creates a new observable each time it produces a `Maximum call stack size exceeded` error.

I fixed that problem by doing this : https://stackblitz.com/edit/angular-ngrx-selectors-good?file=src/app/hello.component.ts

Instead of using selector factories in my template I resolve the whole data structure in a dedicated selector (`selectAllPostsResolved`).

Anyway I think that the workaround is actually better because it remove implementation details from the template, but if this is the recommended way to solve this problem, maybe selector factories are not that useful at all?

**Comment by damienwebdev** [comment]
@timdeschryver @theefer 

We've managed to address this whilst still maintaining memoization (per each `id`).

https://stackblitz.com/edit/ng11-ngrx11-memoized-function-selectors

We actually take memoization further and memoize the factory.

**Comment by theefer** [comment]
Thanks all for the replies, in particular @timdeschryver for the detailed response and the really useful blog post.

You're right that using objects as props will typically break memoization, since object identity will differ. I haven't run any test yet but it might be that we broke memoization for many of our selectors with props when we moved the prop from string to object types. I will confirm but perhaps that proves performance remains OK even in this case.

Do you have examples of the single "viewmodel" selector? I'm not sure exactly what shape you're suggesting - is it combining many other more atomic selectors, or just projecting from larger parts of the state in one selector?

But I like the approach in your blog post to try and move any prop into the state, or at least a selector. This makes total sense and would solve some of our use cases for selector props where they simply represent some ID which is defined by the URL, and could therefore be read using the helper route param selector from ngrx/router.

For the remaining cases where props aren't easily defined by some state/selector (e.g. an enum identifying different collections of items which may or may not occur on the same page, similarly to my example above), I actually realised we could simply memoize the selector factory. It would be similar to what @damienwebdev is suggesting above, but unless I'm mistaken we can do it directly on the outer factory function:

```js
const selectACounter = memoize((id: string) => createSelector(
    selectCounterEntitiesWithLog,
    entities => entities[id]
));
```

where `memoize` is a simple function memoization helper which takes into account the function argument(s). It doesn't need to know anything about ngrx, and it can implement any kind of memoization policy such as custom value comparison (like `defaultMemoize` allows), memory over last N arguments or other caching policy (LRU, MRU, etc), time-based expiry, etc. All it needs to do is to return the same selector, which internally performs its own memoization.

Am I missing something obvious why this wouldn't work? Is there an advantage to @damienwebdev's version?

**Comment by timdeschryver** [comment]
@theefer that should also work, I'm not sure if the behavior is 100% identical though - but it should also provide an answer to your concerns. I'll probably try it out later and compare both methods.

**Comment by Klaster1** [comment]
@theefer the `lodash.memoize` approach works great, it allowed me to update all props selectors to factory selectors and keep existing unit tests that rely heavily on `overrideSelector`. If the selector accepts an object, make sure to pass in the key resolver function as the second `memoize` argument.

**Comment by yzhe554** [comment]
@timdeschryver @brandonroberts 
factory selectors are one solution but I am sure it's more preferred to use the way rereselect do.
And deprecating props will also blocking people using reselect and rereselect....(like us)...

**Comment by sddania** [comment]
@theefer maybe the implementation of `memoize` function is something like this?
```
const memoize = <Result, ArgsType>(filter: (...args: ArgsType[]) => Result) =>
    (...args: ArgsType[]) =>
        defaultMemoize(filter).memoized(args) as Result;
```

**Comment by Nosfistis** [comment]
I think there should be a clearer example of tests migration to factory selectors. Having to use external libraries, such as lodash, just for testing, does not sound like a good idea. Perhaps a utility function could be exposed instead from `@ngrx/store/testing`.

**Comment by lhbruneton** [comment]
I agree some more documentation would be helpful.
In my case, I'm using `provideMockStore` in my unit tests, and since switching to factory selectors, my code builds and runs fine, but my tests are returning with this error: "Cannot read property 'ids' of undefined"

before
```
            provideMockStore({
              selectors: [
                {
                  selector:
                    FaceUpPilesCardsSelectors.getFirstCardPivotForPile,
                  value: {
                    id: 'AAA',
                    pileId: ID_FACE_UP_ROAD,
                    cardId: 'ROAD_1',
                  },
                },
              ],
            }),
```

after
```
            provideMockStore({
              selectors: [
                {
                  selector:
                    FaceUpPilesCardsSelectors.getFirstCardPivotForPile(
                      ID_FACE_UP_ROAD
                    ),
                  value: {
                    id: 'AAA',
                    pileId: ID_FACE_UP_ROAD,
                    cardId: 'ROAD_1',
                  },
                },
              ],
            }),
```

If these factories worked in my unit tests, this would be a huge improvement, as I would now be able to have different return values for different input properties.

**Comment by timdeschryver** [comment]
To test factory selectors, you can take a look at https://github.com/ngrx/platform/issues/3107#issuecomment-892958699

**Comment by maxpontreau** [comment]
Hello,

About discussion between @mmulhearn & @wrslatz about the overrideSelector function.
Indeed it can't be used anymore with those new functions.

But I am still facing issues to mock it with jasmine, maybe because I'm working with angular strict mode.

When I want to mock a selector function like this one below:
```
export const getCurrentScope = (arg: string) => createSelector(getContextState, (state: ContextState) => {
   return state.object + arg;
}
```

well I can't because to mock this function, Angular strict mode is waiting for MemoizedSelector in return.
I did not find a way to perform it, can you give me some ideas plz ?

**Comment by CoderboyPB** [comment]
Don't remove it, maybe offer an alternative solution, but don't force the developers to use a much more complicated system over an easy to use system.

**Comment by progtarek** [comment]
``` 
export const currentUserRoleIds = createSelector(
    currentUser,
    user => {
        if (!user) {
            return [];
        }

        return user.roles;
    }
);

export const currentUserPermissionsIds = createSelector(
    currentUserRoleIds,
    selectAllRoles,
    (userRoleIds: number[], allRoles: Role[]) => {
        const result = getPermissionsIdsFrom(userRoleIds, allRoles);
        return result;
    }
);
```

I was wondering how to update the selector that takes other multiple selectors as an argument, @theefer @mistic100 @brandonroberts @mmulhearn @Klaster1 @timdeschryver

**Comment by Klaster1** [comment]
@progtarek your example selectors do not use props and should work with `overrideSelector`. What's the issue?

**Comment by progtarek** [comment]
@Klaster1 
I am getting this error on starting the app after the upgrade 
**error TS2769: No overload matches this call.** 
![image](https://user-images.githubusercontent.com/18407645/145806461-0c27d98e-f6e9-4d36-9759-8136c9228cfe.png)
![image](https://user-images.githubusercontent.com/18407645/145806625-7a82b83f-24f9-493c-a398-0a16f0165dd0.png)

**Comment by hugo-azevedo** [comment]
@timdeschryver now that selector with props has been deprecated, could you please provide some guidance as to how one is supposed to check the correct arguments are passed to a selector?

Before, this was possible:

```ts
spyOn(store, 'select');
expect(store.select).toHaveBeenCalledWith( selectorWithPropsName, Props );
```

Now, when this is updated to match the new selector syntax

```ts
spyOn(store, 'select');
expect(store.select).toHaveBeenCalledWith( selectorWithPropsName(Props) );
```
we get the test failure message:
`Expected [ Function ] to equal [ Function ].`

Thanks

**Comment by alex-vg** [comment]
There should be a clear alternative given in the deprecation notice.
Currently it's just _"Selectors with props are deprecated, for more info see [[this issue]](https://github.com/ngrx/platform/issues/2980)"._

**Comment by LosD** [comment]
Totally agree with @alex-vg. A feature deprecation should never, ever, be considered without clear and full documentation of the way forward for those using it. Referencing a Github issue does NOT cut it.

@timdeschryver

**Comment by sarancruzer** [comment]
The word "deprecated" in the notice is actually a link, though a hard one to notice due to the color of the banner it's in. It links to [a GitHub Issue discussing selectors with props](https://github.com/ngrx/platform/issues/2980).

That issue raises the concern that, with Angular 12+ being in [strict mode](https://angular.io/guide/strict-mode) by default, developers are likely to run into a shortcoming in the implementation of selectors with props - namely that the prop values are not strongly typed or may be incorrectly typed.

The suggested approach is to use a factory selector. Essentially, instead of passing props as parameters of the selector, you take the props as arguments to a function defined by the selector.


So instead of:

`export const getCount = createSelector(
  getCounterValue,
  (counter, props) => counter * props.multiply
);

const count = this.store.select(getCount, 5);
`

You would use:

`
export const getCount = (multiply: number) => createSelector(
  getCounterValue,
  (counter) => counter * multiply
);

const count = this.store.select(getCount(5));

`


Thnks to https://stackoverflow.com/users/14956277/d-m

**Comment by ezhao1** [comment]
Hello - with the factory selector pattern, it appears we lose the ability to memoize/cache that selector for that specific parameter. Each cycle, we instead get a newly created selector based on the arguments, which then has to re-do its internal logic. In cases where the selector is doing heavy calculation, this can greatly affect performance. What's the recommended way to be able to memoize the calculated value for a factory selector based on the arguments? Would I just have to store the calculated value within the store state itself? (instead of calculating in the selector)

Thanks!

**Comment by labeled** [comment]
Why is this github issue being linked as a deprecated warning on simple selectors? 

![Image](https://github.com/user-attachments/assets/026b9245-e6ba-4774-96b0-16f16fbb97f6)

**Comment by ghost** [comment]
@labeled I get the same issue on non propped selectors

**Comment by dietergeerts** [comment]
> Hello - with the factory selector pattern, it appears we lose the ability to memoize/cache that selector. Each time the function is called, we instead get a newly created selector based on the arguments, which then has to re-do its internal logic. In cases where the selector is doing heavy calculation, this can greatly affect performance. What's the recommended way to be able to memoize the calculated value for a factory selector based on the arguments? Would I just have to store a dictionary mapping the calculated value within the store state itself? (instead of calculating in the selector)
> 
> Thanks!

Besides this, it's also adds extra code to unpack the signal, especially if the input for the props are signals themselves... It would help if there are examples added of how to handle these scenarios.
